# Семинар 2. Домашняя работа 2
## Задание 1

In [ ]:
import pandas as pd

### Считать данные с помощью pandas

In [23]:
df = pd.read_csv('kc_house_data.csv', index_col='id')

#### Вывести на экран первые 5 строк

In [24]:
df.head()

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
7129300520  20141013T000000  221900.0         3       1.00         1180   
6414100192  20141209T000000  538000.0         3       2.25         2570   
5631500400  20150225T000000  180000.0         2       1.00          770   
2487200875  20141209T000000  604000.0         4       3.00         1960   
1954400510  20150218T000000  510000.0         3       2.00         1680   

            sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
id                                                                             
7129300520      5650     1.0           0     0          3      7        1180   
6414100192      7242     2.0           0     0          3      7        2170   
5631500400     10000     1.0           0     0          3      6         770   
2487200875      5000     1.0           0     0          5      7        1050   
1954400510      8080     1.0           0     0          3      8        1680   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
7129300520              0      1955             0    98178  47.5112 -122.257   
6414100192            400      1951          1991    98125  47.7210 -122.319   
5631500400              0      1933             0    98028  47.7379 -122.233   
2487200875            910      1965             0    98136  47.5208 -122.393   
1954400510              0      1987             0    98074  47.6168 -122.045   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503

In [14]:
df.shape

(21613, 20)

#### Изучите типы данных

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21613 entries, 7129300520 to 1523300157
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   date           21613 non-null  object 
 1   price          21613 non-null  float64
 2   bedrooms       21613 non-null  int64  
 3   bathrooms      21613 non-null  float64
 4   sqft_living    21613 non-null  int64  
 5   sqft_lot       21613 non-null  int64  
 6   floors         21613 non-null  float64
 7   waterfront     21613 non-null  int64  
 8   view           21613 non-null  int64  
 9   condition      21613 non-null  int64  
 10  grade          21613 non-null  int64  
 11  sqft_above     21613 non-null  int64  
 12  sqft_basement  21613 non-null  int64  
 13  yr_built       21613 non-null  int64  
 14  yr_renovated   21613 non-null  int64  
 15  zipcode        21613 non-null  int64  
 16  lat            21613 non-null  float64
 17  long           21613 non-null  float

#### Посмотреть на описание признаков и на их содержание

In [16]:
df.dtypes

date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

#### Найдите количество пропущенных ячеек в данных

In [17]:
df.isnull().sum()

date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

#### Посчитайте основные статистики по всем признакам и поизучайте их

In [19]:
df.describe(include=['object'])

date
count             21613
unique              372
top     20140623T000000
freq                142

#### **Выводы:**
#### В данных 21613 строк и 20 столбцов после удаления столбца индекса.
#### Все данные таблицы заполнены.
#### Столбцы и их значение:

- id - Уникальный ID для каждого дома
- date - Дата продажи дома
- price - Стоимость продажи дома
- bedrooms - Кол-во спален
- bathrooms - Кол-во ванных комнат (0.5 - туалет без душа)
- sqft_living - Кв. метры жилые
- sqft_lot - Кв. метры общие
- floors - Кол-во этажей
- waterfront - Есть набержная или нет
- view - Значение от 0 до 4 насколько хороший вид
- condition - Значение от 1 до 5 насколько хорошее состояние
- grade - Значение от 1 до 13, где 1-3 плохая конструкция здания и дизайн, 7 - средний уровень конструкции и дизайна, 11-13 - высокое качество конструкции и дизайна
- sqft_above - Кв. метры дома, которые находятся выше земли
- sqft_basement - Кв. метры дома, которые находятся ниже земли
- yr_built - Год постройки дома
- yr_renovated - Год ремонта дома
- zipcode - Индекс
- lat - Широтма
- long - Долгота
- sqft_living15 - Кв. метры жилой площади у 15 соседей
- sqft_lot15 - Кв. метры общей площади у 15 соседей



## Задание 2
### Ответьте на несколько вопросов
### 3.1 В каком диапазоне изменяются стоимости недвижимости?


In [25]:
df['price'].min(), df['price'].max()

(75000.0, 7700000.0)

### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [71]:
df_sqft_rate = df['sqft_living'].mean() * 100 / df['sqft_lot'].mean()
df_sqft_rate

13.76781757959227

#### 3.3 Как много домов с разными этажами в данных?

In [28]:
df_floors = df['floors'].value_counts()
df_floors

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

#### 3.4 Насколько хорошие состояния у домов в данных?

In [29]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

#### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [45]:
df.sort_values('yr_built').iloc[0, [13]]

yr_built    1900
Name: 9126101645, dtype: object

In [46]:
df.sort_values('yr_built', ascending=False).iloc[0, [13]]

yr_built    2015
Name: 923059259, dtype: object

## Задание 3.
### Ответьте на несколько вопросов
### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?


In [47]:
df[df['bedrooms'] == 2]['price'].mean()

401372.681884058

#### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [49]:
(df[df['price'] > 600000]['sqft_lot'].mean()).round(2)

20442.52

#### 4.3 Как много домов коснулся ремонт?

In [59]:
df_years = df[df['yr_renovated'] != 0]['yr_renovated'].value_counts()
df_years.sum()

914

In [72]:
df[df['yr_renovated'] != 0].shape[0]

914

#### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [61]:
df_top = (df[df['grade'] > 10]['price'].mean()).round(2)
df_top

1678635.12

In [62]:
df_low = (df[df['grade'] < 4]['price'].mean()).round(2)
df_low

189750.0

In [73]:
df_top - df_low

1488885.12

## Задание 4.
### 5.1 Выберите дом клиенту. Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [74]:
df[(df['waterfront'] != 0) & (df['bathrooms'] > 3) & (df['sqft_basement'] != 0)].shape[0]

36

### 5.2 Выберите дом клиенту. Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?


In [83]:
df_selection = ((df['view'] == 4) | (df['waterfront'] == 1)) & (df['condition'] == 5) & (df['yr_built'] >= 1980)
choice = df[df_selection]
choice['price'].min(), choice['price'].max()

(1295000.0, 3000000.0)

### 5.3 Выберите дом клиенту. Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [70]:
(df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] < 150000)]['condition'].mean()).round(3)

2.833